server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1711,13 +1713,180 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
     }
 
     /**
-     * Computes the cluster state resulting from removing a given snapshot create operation from the given state.
+     * Computes the cluster state resulting from removing a given snapshot create operation that was finalized in the repository from the
+     * given state. This method will update the shard generations of snapshots that the given snapshot depended on so that finalizing them
+     * will not cause rolling back to an outdated shard generation.
      *
      * @param state    current cluster state
      * @param snapshot snapshot for which to remove the snapshot operation
      * @return updated cluster state
      */
-    private static ClusterState stateWithoutSnapshot(ClusterState state, Snapshot snapshot) {
+    private static ClusterState stateWithoutSuccessfulSnapshot(ClusterState state, Snapshot snapshot) {
+        // TODO: updating snapshots here leaks their outdated generation files, we should add logic to clean those up and enhance
```

Comment:
The logic in this method would definitely benefit from a state tracking object like we have for the shard status update executor. In the interest of time and keeping it simple I went with this solution for now. Ideally, I'd like to resolve this todo and make the whole logic simpler to follow in 7.15 by finally refactoring `SnapshotsInProgress` into a form that is more appropriate for the logic around concurrent snapshots than to add yet another round of elaborate logic to nicely work around its shortcomings.